### PR TITLE
OGD-370 adding OPTIONS for Individuals API resources

### PR DESCRIPTION
**Pull request by edgarjimenez**
Play does not have direct support for this. In addition:
- It does not like method overloading for routes. There is an added complication to this when using multiple route files.
- It does not like path variables being ignored and you cannot provide a default value for them

The final solution allows the supported methods for a path to be configured in the routes file(s). Configured Http methods are validated using an enumerated type to prevent invalid methods being added by mistake.

### CODE REVIEW

File: conf/app.routes [hunk 1]
```diff
@@ -1,7 +1,12 @@
 # microservice live specific routes
 
 GET         /                            @uk.gov.hmrc.individualsapi.controllers.LiveIndividualsController.hateoas
+OPTIONS     /                            @uk.gov.hmrc.individualsapi.controllers.LiveIndividualsController.options(methods = "GET")
```
Comment (tom59) [inline conf/app.routes]: shouldn't we have an OptionController instead of extending all the existing controllers?

**Comment by tom59** [comment]
+ need to add OPTIONS in the definition so that WSO2 allow it

File: conf/app.routes (hunk 1 above)
Comment (edgarjimenez) [inline conf/app.routes]: could do.

File: app/uk/gov/hmrc/individualsapi/controllers/HttpOptions.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package uk.gov.hmrc.individualsapi.controllers
+
+import org.joda.time.Interval
+import play.api.mvc.{Action, Controller}
+
+import scala.concurrent.Future._
+
+trait HttpOptions extends Controller {
```
Comment (edinhodzic) [inline app/uk/gov/hmrc/individualsapi/controllers/HttpOptions.scala]: I'm not sure it's necessary to extend `Controller` here.

File: test/unit/uk/gov/hmrc/individualsapi/controllers/HttpOptionsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import play.api.test.FakeRequest
+import uk.gov.hmrc.individualsapi.controllers.HttpOptions
+import uk.gov.hmrc.play.test.UnitSpec
+import play.api.http.HeaderNames._
+import play.api.http.Status._
+
+
+class HttpOptionsSpec extends UnitSpec {
+
+  val httpOptions = new TestHttpOptions
```
Comment (edinhodzic) [inline test/unit/uk/gov/hmrc/individualsapi/controllers/HttpOptionsSpec.scala]: To save us defining a new class we could instead have an anonymous class; `new HttpOptions {}`.

File: app/uk/gov/hmrc/individualsapi/controllers/HttpOptions.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  def options(methods: String) = buildResponse(methods)
+
+  def optionsWithMatchId(matchId: String, methods: String) = buildResponse(methods)
+
+  def optionsWithMatchIdAndInterval(matchId: String, interval: Interval, methods: String) = buildResponse(methods)
+}
+
+object HttpMethod extends Enumeration {
```
Comment (edgarjimenez) [inline app/uk/gov/hmrc/individualsapi/controllers/HttpOptions.scala]: we need to validate the configured http methods and we are currently doing that using the enumerated type. HttpVerbs only provides string constants for them, which is not suitable in this instance.

**Pull request merged**